src/symmetric_crypto.cpp

```diff
@@ -0,0 +1,382 @@
+{
+    if (isAuthenticatedCipherMode(_mode)) {
+        throw MoCOCrWException("Specified cipher supports authenticated encryption."
+                               " buildAuthenticatedDecryptor() should be used instead.");
+    }
+    auto cipher = new AESCipher{_mode, _padding, _secretKey, _iv,
+                                AESCipher::Operation::Decryption};
+    return std::shared_ptr<AESCipher>(cipher);
+}
+
+std::shared_ptr<AuthenticatedAESCipher> AESCipherBuilder::buildAuthenticatedEncryptor()
+{
+    if (!isAuthenticatedCipherMode(_mode)) {
+        throw MoCOCrWException("Specified cipher does not support authenticated encryption."
+                               " buildEncryptor() should be used instead.");
+    }
+    auto cipher = new AuthenticatedAESCipher(_mode, _padding, _secretKey, _iv, _authTag,
+                                             AESCipher::Operation::Encryption);
+    return std::shared_ptr<AuthenticatedAESCipher>(cipher);
+}
+
+std::shared_ptr<AuthenticatedAESCipher> AESCipherBuilder::buildAuthenticatedDecryptor()
+{
+    if (!isAuthenticatedCipherMode(_mode)) {
+        throw MoCOCrWException("Specified cipher does not support authenticated encryption."
+                               " buildDecryptor() should be used instead.");
+    }
+    auto cipher = new AuthenticatedAESCipher(_mode, _padding, _secretKey, _iv, _authTag,
+                                             AESCipher::Operation::Decryption);
+    return std::shared_ptr<AuthenticatedAESCipher>(cipher);
+}
+
+
+class QueueOfVectorsMemoryStrategy::Impl {
+public:
+    void write(const std::vector<uint8_t> &block)
+    {
+        _queue.push_back(block);
+    }
+
+    std::vector<uint8_t> read(size_t requestedBufferSize)
+    {
+        if (_queue.empty()) {
+            return std::vector<uint8_t>();
+        }
+
+        if (_queue.front().size() == requestedBufferSize) {
+            return _readCompleteBlock();
+        }
+        else if (_queue.front().size() > requestedBufferSize) {
+            return _readPartialBlock(requestedBufferSize);
+        }
+        else {
+            std::vector<uint8_t> bufferToReturn;
+            bufferToReturn.reserve(requestedBufferSize);
+
+            // First, read as much complete blocks as we can
+            while (!_queue.empty() && (_queue.front().size() <= (requestedBufferSize - bufferToReturn.size()))) {
+                auto block = _readCompleteBlock();
+                std::copy(std::begin(block), std::end(block), std::back_inserter(bufferToReturn));
+            }
+            // Then, read partial block if needed
+            if (bufferToReturn.size() != requestedBufferSize && !_queue.empty()) {
+                auto block = _readPartialBlock(requestedBufferSize - bufferToReturn.size());
+                std::copy(std::begin(block), std::end(block), std::back_inserter(bufferToReturn));
+            }
+
+            return bufferToReturn;
```

Comment:
Please move this buffer into the output. The same applies for the other functions returning vectors in this class, but I haven't commented on each occurrence separately.

Reply:
wouldn't it be RVOed?